jopa-impl/src/main/java/cz/cvut/kbss/jopa/sessions/ReadOnlyUnitOfWork.java

```diff
@@ -0,0 +1,434 @@
+        }
+    }
+
+    private void registerExistingObjects(Iterable<Object> collection, Descriptor descriptor) {
+        for (Object entity : collection) {
+            if (!super.isEntityType(entity.getClass()) || isObjectManaged(entity)) { return; }
+            if (isObjectInCache(entity.getClass(), super.getIdentifier(entity), descriptor)) {
+                registerExistingObject(entity, new CloneRegistrationDescriptor(descriptor)
+                        .postCloneHandlers(List.of(new PostLoadInvoker(getMetamodel()))));
+            } else {
+                registerExistingObject(entity, descriptor);
+            }
+        }
+    }
+
+    @Override
+    public Object getOriginal(Object original) {
+        // simply return the original object
+        // TODO: change javadoc
+        return original;
+    }
+
+    @Override
+    boolean containsOriginal(Object entity) {
+        assert entity != null;
+        return originalMapping.contains(entity);
+    }
+
+    @Override
+    public boolean isObjectManaged(Object entity) {
+        Objects.requireNonNull(entity);
+        return this.originalMapping.contains(entity);
+    }
+
+    @Override
+    public EntityState getState(Object entity) {
+        Objects.requireNonNull(entity);
+        return originalMapping.contains(entity) ? EntityState.MANAGED : EntityState.NOT_MANAGED;
+    }
+
+    @Override
+    public EntityState getState(Object entity, Descriptor descriptor) {
+        Objects.requireNonNull(entity);
+        Objects.requireNonNull(descriptor);
+
+         return originalMapping.contains(entity) && super.isInRepository(descriptor, entity)
+                 ? EntityState.MANAGED
+                 : EntityState.NOT_MANAGED;
+    }
+
+    @Override
+    public <T> T getReference(Class<T> cls, Object identifier, Descriptor descriptor) {
+        return super.readObject(cls, identifier, descriptor);
+    }
+
+    @Override
+    public Object getCloneForOriginal(Object original) {
+        // this unit of work does not track clone-original
+        // simply return the original object
+        return original;
+    }
+
+    @Override
+    public Object createIndirectCollection(Object collection, Object owner, Field field) {
+        // Do not create any special kind of collection, just return the argument
+        return collection;
+    }
+
+    private boolean isObjectInCache(Class<?> cls, Object identifier, Descriptor descriptor) {
+        return getLiveObjectCache().contains(cls, identifier, descriptor);
+    }
+
+    @Override
+    public void putObjectIntoCache(Object identifier, Object entity, Descriptor descriptor) {
+        // object should never be put into cached in this uow
+    }
+
+    //////////////////////////////////////THESE METHODS SHOULD NOT BE SUPPORTED/////////////////////////////////////////
+    @Override
+    public void removeObjectFromCache(Object toRemove, URI context) {
+        throw new UnsupportedOperationException("Method not supported.");
```

Comment:
Extract the throw into a static function and call it in all the relevant places instead of repeating the same constructor call with the same string literal.